pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -545,29 +561,23 @@ async fn loop_step<
     new_timeline_states
 }
 
+// this is where new TimelineState borns
 async fn process_task<
     P: std::fmt::Debug + Send + Sync + 'static,
     S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
 >(
     conf: &'static PageServerConf,
-    remote_assets: Arc<(S, RwLock<RemoteTimelineIndex>)>,
+    remote_assets: Arc<(S, Arc<RwLock<RemoteTimelineIndex>>)>,
     task: SyncTask,
     max_sync_errors: NonZeroU32,
-) -> Option<TimelineSyncState> {
+) -> Option<TimelineSyncStatusUpdate> {
     if task.retries > max_sync_errors.get() {
         error!(
             "Evicting task {:?} that failed {} times, exceeding the error threshold",
             task.kind, task.retries
         );
-        return Some(TimelineSyncState::Evicted(
-            remote_assets
-                .as_ref()
-                .1
-                .read()
-                .await
-                .timeline_entry(&task.sync_id)
-                .and_then(TimelineIndexEntry::disk_consistent_lsn),
-        ));
+        FATAL_TASK_FAILURES.inc();
+        return None;
```

Comment:
If we ignore an upload, we would never be able to retrieve that lost checkpoint file list again, AFAIK.
We would allow to continue uploading further checkpoints for such timelines, which might lead to us having timelines with "holes" we never detect.

I would disallow further uploads for now, to make it more explicit? And hold such "evicted" status in the index, maybe?

Reply:
What do you think about rescheduling instead of ignoring and an additional check for uploads, so there is no upload with smaller lsn in the queue?

Reply:
That could bloat our queue with tasks that will always fail, couldn't it?
> so there is no upload with smaller lsn in the queue

Such check IMO also a bit dangerous, since someone could implement the checkpointing diffrently.
This also worsens the "constant fails" scenario in case new checkpoints will arrive after the problematic one.

I have no good solution to the issue, shutting down timeline uploads entirely does not seem right either.

------
We might allow uploading checkpoints in any order, yet keep track of "holes" and the files that were lost: then we can easily track such timelines and never touch their metadata files (yet allowing to sync other files)

Reply:
> That could bloat our queue with tasks that will always fail, couldn't it?

Thats true, but I do not see a better option

> This also worsens the "constant fails" 

Could you elaborate on that? I see that there would be more failures because of the incorrect order indeed

> shutting down timeline uploads entirely does not seem right either 

Yeah, I think that's even worse than having a big queue. I think in that case backpressure should kick in because if checkpoints are not uploaded then we do not advance remote consistent lsn. We cannot stop checkpointing either. Maybe we can introduce some limit for number of checkpoints that are in progress of uploading, not sure

Reply:
> Could you elaborate on that?

Just thought that with the 

> an additional check for uploads, so there is no upload with smaller lsn in the queue

approach we're either making too many gaps in the remote checkpoint collection or forced to reschedule those tasks, they will accumulate and task processing will degrade over time, since required to go through more and more tasks, checking.

Either way, no good solution now is fine, we don't have to force it now, let's merge the test fix.

Reply:
Mentioned this in the #977 plan to work on that in a follow up